src/tree/updater_quantile_hist.cc

```diff
@@ -740,19 +761,34 @@ void QuantileHistMaker::Builder<GradientSumT>::InitSampling(const std::vector<Gr
       const size_t ibegin = tid * discard_size;
       const size_t iend = (tid == (nthread - 1)) ?
                           info.num_row_ : ibegin + discard_size;
-
-      rnds[tid].discard(discard_size * tid);
+      constexpr uint64_t kBase = 16807;
+      uint64_t initial_seed_th = SimpleSkip(discard_size * tid, initial_seed, kBase);
+      std::linear_congruential_engine<std::uint_fast64_t,
+                                      kBase,
+                                      0,
+                                      static_cast<uint64_t>(1) << 63> eng(initial_seed_th);
+      size_t local_unused_offset = 0;
+      size_t local_used_offset = 0;
       for (size_t i = ibegin; i < iend; ++i) {
-        if (gpair[i].GetHess() >= 0.0f && rnds[tid]() < coin_flip_border) {
-          p_row_indices_used[ibegin + row_offsets_used[tid]++] = i;
-        } else {
-          p_row_indices_unused[ibegin + row_offsets_unused[tid]++] = i;
+        if (!(gpair[i].GetHess() >= 0.0f && coin_flip(eng)) || gpair[i].GetGrad() == 0.0f) {
+          p_row_indices_unused[ibegin + local_unused_offset++] = i;
+          if (is_single_tree) {
```

Comment:
Why is single tree different?

Reply:
for single tree we can change (set to zeros for unused rows) initial gradient vector while for not single tree case we have to reset initial values for each tree assigned for iteration (special case `num_parallel_tree != 1`). So copy is done only for not single tree case and removed this condition from `InitSampling`.